### PR TITLE
Catch divide by zero in GMP rational assignments

### PR DESCRIPTION
**Pull request by ckormanyos**
This is intended to handle #451

Hi John (@jzmaddock) here is a first cut attemting to catch div-by-zero in GMP rational assignments/ctors.

- Could you please take a look if I found all (and the right) places?
- Also I was a bit concerned for handling the move assignments properly, and I'm unsure if you agree with what i did?
- I wonder if we should add some explicit GMP-only tests in the arithmetic-tests? Such that these try and catch div by zero?

I also handled that single remaining MSVC Level 4 warning here (although it's a bit off topic).

### CODE REVIEW

File: include/boost/multiprecision/gmp.hpp [hunk 1]
```diff
@@ -2380,6 +2384,9 @@ struct gmp_rational
    template <class T, class U>
    gmp_rational(const T& a, const U& b, typename std::enable_if<std::is_constructible<gmp_int, T>::value && std::is_constructible<gmp_int, U>::value>::type* = nullptr)
    {
+      if (b == static_cast<U>(0))
```
Comment (jzmaddock) [inline include/boost/multiprecision/gmp.hpp:2387]: I don't think we have tests that would catch this, but I would worry about T and U being types which gmp_int is constructible from, but which do not have an `operator=(int)`, an example would be `mpz_t`.

So I would be inclined to put this check after the construction of the gmp_int's below and use a call to eval_is_zero for the check.  Hope that makes sense!

File: include/boost/multiprecision/gmp.hpp [hunk 2]
```diff
@@ -2390,6 +2397,9 @@ struct gmp_rational
    template <class U>
    gmp_rational(const gmp_int& a, const U& b, typename std::enable_if<std::is_constructible<gmp_int, U>::value>::type* = nullptr)
    {
+      if (b == static_cast<U>(0))
```
Comment (jzmaddock) [inline include/boost/multiprecision/gmp.hpp:2400]: same here, and in a few places below.

**Comment by jzmaddock** [comment]
There's a couple of comments attached to source, but I'll work up some generic tests to go in test_arithmetic.hpp and then we can see what breaks ;)

File: include/boost/multiprecision/gmp.hpp [hunk 3]
```diff
@@ -1753,6 +1753,10 @@ inline typename std::enable_if<boost::multiprecision::detail::is_arithmetic<T>::
    return a.compare(b) > 0;
 }
 
+inline bool eval_is_zero_move(gmp_int&& val)
```
[the remark this answers is not in the document]
Comment (ckormanyos) [reply include/boost/multiprecision/gmp.hpp:1756]: Agreed. Thanks John. I was unsure of that conversion.

File: include/boost/multiprecision/gmp.hpp (hunk 1 above)
Comment (ckormanyos) [reply include/boost/multiprecision/gmp.hpp:2387]: I understand. This makes sense. i will tweak these and similar ones accordingly.

**Comment by jzmaddock** [comment]
Tests are coming.... there are more failures with cpp_rational/rational_adaptor than I was expecting :(

**Comment by ckormanyos** [comment]
> Tests are coming.... there are more failures with cpp_rational/rational_adaptor than I was expecting

Understood.

Sorry John, I think I might have moved too quickly forward on merging. I believe we will need a new PR for adding the tests...

I will wait on any/all further activities until we resolve this phenomenon.
